db/snapshot/backup_task.cc

```diff
@@ -99,52 +101,84 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client>& client, sstr
+                backup_files.emplace_back(size, std::move(component_ent->name));
             }
-        }
-        if (!component_ent.has_value()) {
+        } catch (...) {
+            ex = std::current_exception();
             break;
         }
-        auto gh = uploads.hold();
+    } while (component_ent);
+    co_await snapshot_dir_lister.close();
+    if (ex)
+        co_await coroutine::return_exception_ptr(std::move(ex));
 
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    std::ranges::sort(backup_files, std::greater());
+    co_return backup_files;
+}
+} // namespace
 
-        std::ignore = upload_component(cln, component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
-        co_await coroutine::return_exception_ptr(std::move(ex));
+    auto backup_files = co_await get_backup_files(_snapshot_dir);
+    std::atomic_size_t counter = 0;
+
+    sharded<abort_source> as;
+    co_await as.start();
+    auto s = _as.subscribe([&]() noexcept {
+        std::ignore = smp::invoke_on_all([&as, ex = _as.abort_requested_exception_ptr()] {
+            as.local().request_abort_ex(ex);
+        });
+    });
+
+    co_await smp::invoke_on_all([this, &backup_files, &counter, &as] -> future<> {
+        auto shard_id = this_shard_id();
+        gate uploads;
+
+        try {
+            auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+
+            size_t name_idx = counter++;
+            while (name_idx < backup_files.size()) {
+                auto gh = uploads.hold();
+                // Pre-upload break point. For testing abort in actual s3 client usage.
+                co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+
+                std::ignore = upload_component(cln, as.local(), _progress_per_shard[shard_id], std::get<1>(backup_files[name_idx])).handle_exception([this](std::exception_ptr e) {
+                    _as.request_abort_ex(std::move(e));
```

Comment:
Another std::ignore. Don't ignore things that were meant not to be ignored without an explanation.